keps/sig-cluster-lifecycle/kubeadm/20191125-remove-clusterstatus-from-kubeadm-config.md

```diff
@@ -0,0 +1,138 @@
+- [x] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP is proposing a new mode for tracking the list of the API endpoints in a cluster, thus allowing to remove the  `ClusterStatus` entry in the `kubeadm-config` ConfigMap and solve the problems that arise when, for any reasons, such entry does not reflect anymore the real status of the cluster.
+
+## Motivation
+
+In order to manage HA cluster properly, kubeadm requires to have access to the list of API endpoints in a cluster.
+
+Currently this feature is implemented by adding an entry in the list of API endpoints that is stored in the `ClusterStatus` entry of the `kubeadm-config` ConfigMap.
+
+There are well known problem related to the management of this list e.g. when a control-plane node dies or is deleted without invoking `kubeadm reset`, the list gets stale and the user is required to manually cleanup the list in order to avoid any kubeadm operation that relies on such list might incur into errors.
+
+This KEP is going to propose a different mode for tracking the list of the API endpoints in a cluster, based on the inspection of the current Pods.
+
+This approach does not require the maintenance of a separated list, and implicitly always reflect the current status of the cluster.
+
+This allows to remove the `ClusterStatus` entry in the `kubeadm-config` ConfigMap and to clean-up all the related code in `kubeadm init`, `kubeadm join` and `kubeadm reset`
+
+### Goals
+
+- To introduce a new method for tracking the list of API endpoints in a cluster
+- To allow removal of the `ClusterStatus` entry in the `kubeadm-config` ConfigMap and clean-up of the related goal.
+
+### Non-Goals
+
+- To change any user facing behavior in `kubeadm`
```

Comment:
we still need to implement the change as "action required". someone might depend on the ClusterStatus even if we might have not wanted them to.

Reply:
This is our case, for instance, so I could expect other users reading this information as well.

Reply:
I personally think that CLI and API are clearly regulated in terms of deprecation policy, but IMO considering the `kubeadm-config` ConfigMap on the same level or considering this as a user-facing behavior is a stretch. This is a kubeadm internal.

Said that I'm fully supportive of the action required/anything else that can help users, but considering that the current ClusterStatus can easily become stale, we should consider if continue to support users reading from this ConfigMap is a good thing or self-shooting on our feet.

Reply:
done!